replica/table.cc

```diff
@@ -2057,9 +2082,12 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
             table_id, new_tablet_count, old_tablet_count * split_size));
     }
 
+    // Even though the loop over storage groups does not erase map's elements, it stops their components
+    // and so we need to hold exclusive lock throughout the whole process.
+    rwlock::holder exclusive_lock = co_await get_rwlock().hold_write_lock();
     // Stop the released main compaction groups asynchronously
```

Comment:
I am thinking we cannot yield here after tablet map was already updated in tablet_storage_group_manager::update_effective_replication_map(), as storage groups must reflect the new tablet map.

We may have to put this exclusive lock in the beginning of table::update_effective_replication_map(), before we set the new erm. The lock could also be helpful to protect against successive changes to erm (one arriving right after the other) which affect a particular shard. So changes to storage structure are serialized.

@tgrabiec Can you think of any problem if we yield before setting the new erm? Different shards might arrive at new erm at different times, but that's already an issue today if shard 0 is responsible, while shard 1 is busy with a stalling task. We only need to make sure that the replica storage will always reflect the current erm, right? If a request arrives at shard 1 while it wasn't updated with new token metadata yet, it should still be able to map a token at a particular storage group.